artemis-protocols/artemis-openwire-protocol/src/main/java/org/apache/activemq/artemis/core/protocol/openwire/amq/AMQSession.java

```diff
@@ -302,108 +301,91 @@ public void send(final ProducerInfo producerInfo,
-                             ServerMessage originalCoreMsg,
-                             final Runnable onComplete) throws Exception {
+      final AtomicInteger count = new AtomicInteger(actualDestinations.length);
 
-      Runnable runToUse;
+      final Exception[] anyException = new Exception[] {null};
 
-      if (actualDestinations.length <= 1 || onComplete == null) {
-         // if onComplete is null, this will be null ;)
-         runToUse = onComplete;
-      } else {
-         final AtomicInteger count = new AtomicInteger(actualDestinations.length);
-         runToUse = new Runnable() {
-            @Override
-            public void run() {
-               if (count.decrementAndGet() == 0) {
-                  onComplete.run();
-               }
-            }
-         };
+      if (shouldBlockProducer) {
+         connection.getContext().setDontSendReponse(true);
       }
 
-      SimpleString[] addresses = new SimpleString[actualDestinations.length];
-      PagingStore[] pagingStores = new PagingStore[actualDestinations.length];
-
-      // We fillup addresses, pagingStores and we will throw failure if that's the case
       for (int i = 0; i < actualDestinations.length; i++) {
          ActiveMQDestination dest = actualDestinations[i];
-         addresses[i] = new SimpleString(dest.getPhysicalName());
-         pagingStores[i] = server.getPagingManager().getPageStore(addresses[i]);
-         if (pagingStores[i].getAddressFullMessagePolicy() == AddressFullMessagePolicy.FAIL && pagingStores[i].isFull()) {
-            throw new ResourceAllocationException("Queue is full");
-         }
-      }
-
-      for (int i = 0; i < actualDestinations.length; i++) {
+         SimpleString address = new SimpleString(dest.getPhysicalName());
+         PagingStore store = server.getPagingManager().getPageStore(address);
 
          ServerMessage coreMsg = originalCoreMsg.copy();
+         coreMsg.setAddress(address);
 
-         coreMsg.setAddress(addresses[i]);
-
-         PagingStore store = pagingStores[i];
-
-         if (store.isFull()) {
-            connection.getTransportConnection().setAutoRead(false);
-         }
+         if (shouldBlockProducer) {
 
-         if (actualDestinations[i].isQueue()) {
-            checkAutoCreateQueue(new SimpleString(actualDestinations[i].getPhysicalName()), actualDestinations[i].isTemporary());
-         }
+            if (!store.checkMemory(() -> {
+               try {
+                  RoutingStatus result = getCoreSession().send(coreMsg, false, dest.isTemporary());
 
-         if (actualDestinations[i].isQueue()) {
-            coreMsg.putByteProperty(org.apache.activemq.artemis.api.core.Message.HDR_ROUTING_TYPE, RoutingType.ANYCAST.getType());
+                  if (result == RoutingStatus.NO_BINDINGS && dest.isQueue()) {
+                     throw new InvalidDestinationException("Cannot publish to a non-existent Destination: " + dest);
```

Comment:
you are throwing the exception inside the checkSize callback, which could eventually happen outside of the client's thread. (You are sending a Runnable to checkSize).